apps/docs/pages/guides/auth/auth-postgres-hooks.mdx

```diff
@@ -0,0 +1,614 @@
+    - `http_code` A number indicating the HTTP code to be returned. If not set, the code is HTTP 500 Internal Server Error.
+    - `message` A message to be returned in the HTTP response. Must be set.
+
+You should return an error if an input in the `event` payload has been processed incorrectly. This error could result from a server error (e.g. An unexpected error with the data) or a user error (e.g. the user has exceeded the number of available attempts).
+
+In general, Supabase Auth will ignore all fields other than the `error` field attached to the returned JSON.
+
+Example output:
+```jsx
+{
+  "error": {
+    "http_code": 429,
+    "message": "You can only verify a factor once very 10 seconds."
+  }
+}
+```
+
+### Assigning Hook Permissions
+
+You will need to assign additional permissions to the hook as well as the tables that it will interact with.
+
+By default, Supabase Auth will invoke a hook with the `supabase_auth_admin` role and the `supabase_auth_admin` role does not have permissions to modify the `public` schema - you will need to grant the role additional permissions to execute the function you have defined:
+
+```sql
+grant execute
+	on function public.custom_access_token_hook
+	to supabase_auth_admin;
+```
+
+Also revoke permissions from the `authenticated` and `anon` roles to ensure the function is not accessible by Supabase Serverless APIs.
+
+```sql
+revoke execute
+	on function public.custom_access_token_hook
+	from authenticated, anon;
+```
+
+<Admonition type="warning">
+Avoid adding a `security definer` tag to the function to avoid permission errors.
+</Admonition>
+
+At time of writing, Postgres functions have an implicit tag of `security invoker` and will run with the permissions of the user that calls it. In the case of Auth Hooks, `supabase_auth_admin`  is the invoker and it will run with the permissions of `supabase_auth_admin`.
+
+The `security definer` tag specifies that the hook is to be executed with the privileges of the user that owns it.  In the case of functions created via the Supabase Dashboard this is `postgres` which does not have permissions over the Auth schema. Introducing a `security definer` tag could prevent Supabase Auth from having the appropriate permissions to access tables or execute functions.
+
+
+### Timeouts
+
+Hook timeouts are controlled by a `local` statement timeout scoped to the Postgres Function Hook. Ensure that your hooks complete within **2 seconds** to avoid any errors
+
+### Testing the Hook
+
+We strongly recommend testing the hook prior to going into production via [PgTAP](https://supabase.com/docs/guides/database/extensions/pgtap) or similar. You can enable the extension under `Database > Extensions > PgTAP`
+
+Here's an example of what a `PgTAP` test could look like for an MFA Verification hook.
+
+```sql
+begin;
+select plan (1);
+
+-- table to store test results
+create temporary table test_results (
+  name text primary key
+  , result boolean
+  , debug text
```

Comment:
```suggestion
  name text primary key,
  result boolean,
  debug text
```

we style all other sql blocks in docs with trailing commas so let's just that for consistency